cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,370 @@
+  for (const auto [label, P] : label_Ps[node_i_id]) {
+    new_label_Ps_i.insert({label, w_selfloop / sum_w[node_i_id] * P});
+  }
+
+  // propagate neighbors’ probabilities
+  for (const auto node_j_id : InNeighborsMemgraphIDs(node_i_id)) {
+    double contribution =
+        GetTotalWeightBetween(node_j_id, node_i_id) / sum_w[node_i_id];
+    for (const auto [label, P] : label_Ps[node_j_id]) {
+      new_label_Ps_i[label] += contribution * P;
+    }
+  }
+
+  return new_label_Ps_i;
+}
+
+void LabelRankT::Inflate(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps, double exponent) {
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
+    node_label_Ps[label] = inflated_node_label_Ps;
+    sum_Ps += inflated_node_label_Ps;
+  }
+
+  for (const auto [label, _] : node_label_Ps) {
+    node_label_Ps[label] /= sum_Ps;
+  }
+}
+
+void LabelRankT::Cutoff(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    double min_value) {
+  std::vector<std::uint64_t> to_be_removed;
+
+  for (const auto [label, P] : node_label_Ps) {
+    if (P < min_value) to_be_removed.push_back(label);
+  }
+
+  for (const auto label : to_be_removed) {
+    node_label_Ps.erase(label);
+  }
+}
+
+std::pair<bool, std::uint64_t> LabelRankT::Iteration(
+    bool incremental, std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete) {
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      updated_label_Ps;
+
+  bool none_updated = true;
+  std::uint64_t most_updates = 0;
+
+  for (const auto node_id : NodesMemgraphIDs()) {
+    if (incremental) {
+      bool was_updated = changed_nodes.count(node_id) > 0 ? true : false;
+      bool was_deleted = to_delete.count(node_id) > 0 ? true : false;
+
+      if (!was_updated || was_deleted) continue;
+    }
+
+    // node selection (a.k.a. conditional update)
+    if (!DistinctEnough(node_id, similarity_threshold)) continue;
+    none_updated = false;
+
+    // label propagation
+    std::unordered_map<std::uint64_t, double> updated_node_label_Ps =
+        Propagate(node_id);
+
+    // inflation
+    Inflate(updated_node_label_Ps, exponent);
+
+    // cutoff
+    Cutoff(updated_node_label_Ps, min_value);
+
+    updated_label_Ps.insert({node_id, updated_node_label_Ps});
+  }
+
+  for (const auto [node, updated_node_label_Ps] : updated_label_Ps) {
+    label_Ps[node] = updated_node_label_Ps;
+    times_updated[node]++;
+
+    if (times_updated[node] > most_updates) most_updates = times_updated[node];
+  }
+
+  return {none_updated, most_updates};
+}
+/* #endregion */
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::CalculateLabels(
+    std::unique_ptr<mg_graph::Graph<>>& graph,
+    std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete, bool persist) {
+  this->graph = std::move(graph);
```

Comment:
This move is really sneaky. I would use `std::unique_ptr<mg_graph::Graph<>>&&` as parameter type, therefore the callers are forced to move the unique pointer as kind of explicitly handling over the ownership.

Reply:
@antepusic asked me for an explanation for this in private, so I will leave [this](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYgATKVpMGoZAFI9AIQuXSS%2BsgJ4BlRugDCqWgFcWDEJIA7KRuADJ4DJgAcr4ARpjEElwArKQADqgKhM4Mnj5%2BAcEZWU4C4ZExLPGJXCl2mA6lDEIETMQEeb7%2BQfWNOS1tBOXRcQlJqQqt7Z0FPZODw5XV4wCUdqjexMjsHBYAzBHIPlgA1GZ77s6TxJis59hmGgCCj08A9G8nACoIeAonAHc6LQTvETkwTgA3BIAT1BxDwmCoJwiBGIGG8jhyJ1QyJYqGhJyULGMTmQClIoMwom8ShArw%2BJxOXAAdN8EAlMOCbicWJiEFSCAQEidMKo0kZmE1/hsCCcCByeQrjCcAJInNDeWjoUXQ4iAhHC0gMz5MpmYFnAFkgE4IIVpBQgD7zZAAawJCSotFQAJZaBYbyYbz2XA0AE4wwAOLhvABssb2EcjYZNTL0bPV6AEYA4coBpPlqANhC5ClQbAVEWAJyzAIYKPrWFadH%2BeGRZbYsQwcIaSnBtBuTHQMNTZpreB1hBZqcs3jl6tYtqYaTScKIJ2AeEJEIUCBIctR6PQmKa8oQTAP/yYsTLPmFJ0i2wUCjaa9QqeMCgBIoBCDhaSYZ8qzPLlvAYPAAEdvEwAB9NI0VBOEzFjDRJnQEAQDAyDoLgtFzncABxYhlwQfD7nIvRYwsWMUX%2BeIhQSVMIjPP5RVUJhHCJQhvAvHJp2eRkfi5fFCWJUk8HJUVmFieh/nXSZbhBBUuRuMtNifHFkRVVBlP1VBYm0akCDZABZJgYXiVNlN5D0NQEa4TxIDUVTBKhWDoPA2kLE4YQ2elnleBzOKEM5AmsZ4xyECAVhOG1UVw4gzDCyIAQbAgIA0FYkoAEVCywiQIdCUFlM4Lnws49D0LA3K1OVHBIMxkncBgLD0c58py15IogNAGEmE4hGonFdJiuLBASpLLBStKIAAKh0xUWXi%2BDiCywJcsmrqxzNNCMM1OUyIuCq9DQVcNSIRKmpayr2q2plOois0osGyihoWhJRrSiawve4glvGlacryu7tt%2B/6CAS0rcoYLVaEBvZwqebadsKvaSsO9xjpErl6su5rWtux77vWkGAD8opizaie2tsTggGHaDhhDzmylmvpWymwpB5GmV24q51K9xyta9ALsBQg9wFrB6GFRr8ZuhHXm5s0gd7LkqaRnmxz5/bBeFyrRacoEFRrBpMFlq6CcV6mtel82uWWvDrc15GHpd0LWYC6muy8E4AAlAIANTECdlDRaK7L6g6wpOG4CE2esGaZ/UwDANnHcShGPZB1ETgI83g58TAI96/rJtj82E5OWaM/a7PqchVAJwGguxGgiBc8hNvME5/Ka4B5m9lyrui7rt27rSBEu%2BFfz3dRWb2adxHOud1N1Urf4jkAuSOS44ApU2LlcRswlcdogaIBeqjKJWNkoh9E4ZZzOSxFdcEuwF38L1TaihrYT9U5qjlP/Pq4J6zEBHtBCuVBOQMG2GyAA6hLEqwAm4mFNpgNIaUizWUgZgVMohhSoIRJgVs9YhZWBsGcZIlguCNVZjQtqyRcpPw4P8F8a5d4QO7hXSepDGAECvAwHU%2BlDKcWVAQVMqkNhbFIc5esYJJheEYIhYaiocQGSMmyISsc8HQNgRpNowlMgEFoCOASnxkCDmFDqWIcJaTAT5tjahlg9j0MpL%2BCSApWLaFpHKPUcIFAknaM5fq65uGjwsfom4cCuTrgiPgQhcTzxym0poxwz8K5qVkZvFypYiD0AYPxF4USnhUGFPqDeHicYqmceuGRkcrH2zYqwCUFolbPEbs3IQeAWBtIAEpIkvkNBQvcQaCV%2BK2f4X9LxLhXIwYCDAfTaMmefIgrRGZwhHs3Ch1grCUliALQgtFxmfHQMYYACQNj/CoOiFgIEiS4gIPmHkWAsj7xOGkBICgvlYmhHffS3Y2K/MEWeb%2B1NGTWWkepORx9kKoXhdyLkqDgLrkUQUlRdiwG/Q0WI4y7J8EQs%2BGiKBxsBQQiwlBWCK1KROUwNCes%2BIbgrIvJk38cJjmsSWXKVBGBCynM%2BdvbkGxhEnEpThFa0TGBPhtLSUsvS2mfLQRUmUiUiVStiTcpy1ksh9PoKE0htL9Ryt0Twm4MCYkaV/Co5SAroWyM0mK8CVLIYsDwMAO0VICrKIUUhSqIlgK3PLK1FExkQZMDnEWb0ohaAwX%2BGzJxHoICjMJi8EmXsnhdJ1P0wu0FBlUGGa9SiRIxnqp%2BKxKgYEsQCEfqgVArp/i%2BP6rQPArp1YoR6XqzA%2Bb4UHKOXKG42wtxyPhQoRFdi7V6PNQYw1D4dLgi4l2jV2w6a0h4ps5yjNMA6loHgu1SIZ233ZKxEB8aUKJuhIi88ckixgjlTqa19YIiN1dFWAV1lK1wKaGyJ4orvxztSqO1GIBsaIoSRJC8ciJEPNEUZZ%2B%2B7skaTdR6vMwIvVKMKSyggmSVRiglBJY5/o3UEDYIIAVQb7nWRjfQfUWQsDFLHBG9c0axBxqhgVIq2Nk1ZVXumkp7xPihHNpkpQXI9ypX/aIJ9DAX3JMwCwVOrxc4kgiBHDWY51M8x1ujMqR1WqdoGUiOW102rOy1iFMdZmtPAd1hjY6CgWQBwUIXUO4cVg2hDXZhzTmXPoDDsQNTunMY5mMyFqzyNtMCzs61dwYgW3oIM/QHtlsFaIy1oyJ4/xfLeHkUSTAYnORzsrOg1iPzqSeVoH2g657x3UnLOrSq8n4IjkquCcpjEbZmkZNR4C8LEvdqMyhBjPN%2Bv5u46mrWkWDpBeOqNpEFd47EEiOgYzVs0vWaKrZmbrVvNBxDn5tzHnWtecc3tltB2AuUxmyFq6YX1vE09u7Kbes9M3UtiltxehPvff2L9r7f2fv/aB4DkHAOwfvflqZ%2B7wNOu8xszpoWr29A5u7slyHE2RtEgxxF%2BHUXtuVV285/b/nopHb0C9zGhPfMk6u4j4Luxbu7HC9tZ70XKqxcZsBFHRc0cmex9tRkftCv/rcGKrBzEdWsHlfHXiNabjADaPgdB65R29tDeLRmVJlZMkZE1ggcImDtf1PC7nebBsaCq4CDk9YOyYC7MOQEpIb0Nlk6G40sOTiMiA5xj0V7ALeTvUoHU65qPq7JWA/sg4Hdij%2BE4Ew2vPefFg5xRqlgmD0LZJYOE/Ati9fPcB0DKEHnUZ/GhsETAXnJIvGcGrKEE%2BMmQ56sEevEQ6kNxUzd8Xqwfqrd%2Bk4v7g%2BrNYgq/cpJ3fuzHIyeINI%2BwOPQWkvF0zfjIAFNeqkKjsY6go%2BfNdcWDd2SwPXz4QSGggmMDqVStUyEak2M4WkoJvDAAcwn03A2C0XuLqMnj0PJ%2BfFf2NlNLORkbKVAUhBgHMOqcsNIOgTAA5akCNPsf9c8QkdcE1b3DCQvDQBPVnfHZHXNN/BbBObdVbVLbA3HabOnezU7Inc7GnMnCnIkag6nNzBgm7ZqO7EGcePjVMTLBdS/WgecFiBgN%2BD9Lwb0IEdBFfC5ORXcDYbUL1EkU4CXXeNALAWIQCQlATCuBXYgJXbvIseFDnZALUSDUIa8XsekOvKJVkE4dwc8EwOTeUGEL5R1YAYiNIAUbfXdM1A9S1LkCAaiGKcJPdKJM0adPwumX%2BG%2BYpRkdMfvdAHUdAkDX3Iva1HkQwsQYwowYUMw%2BIWgMdIvViEvdAGIz4PYH9K8LJWqHyDYNKI8TEbdGvVCAvFIjQbyHFT9atPqSkXSJEEgTQxkM0IcJtOUayACYiCsEUfXFw4%2BBUTILkTo6UMBNvBIpoj/RFdcZSPANVLQs0Z9VAQhPiVMSQNkQZL5SDQfLkPhSEO/f4DIVEf4MCJwEEbLCuDiAUHvL9bENIxwtwkiUcJkIo6xbdLDE4RY7EayP4jw8%2BblCuIdaENvSo8YqXCpKrAE9XIE24GxWtajGESkMsGonLIxMEm4OJIsOpXeZPVDOYucdEqQkwYCYwGEYrasNWBsMFOUTowgAQYpH%2BGhdPZhb4akBAcCXE9XKTfsAksEYo8WE2YUPpEgV8XFIyCkUUS0a0H%2BFCf/IZQaLmdaLKFCOAmfOJJrRU4gOEKkq8HkCJaCOiOEaqCNAQnohAe/CIKyVZUQPsJIzAmE%2BdR8UhF8c03kqJVPOhZhVMO0AgB0J0N4LMckFkN1KxTIJ5P0csN4RgAAWlpDeGQBXBzLzO8KLgUAzPPwzJtNIQzJuIUHXQzNzLSHOAADEbjMAAQWYWAFBIRkAMyuArCtDU8mFcpIzoyPhGA/QVxwjpULR/Q3gAR8y0g3hJRgAeJLk3g8EYIkliFzE%2ByaE3FBShzHQRyik6yJzYlUyAxZy6y3g5wPJ9c3hsYOA1haBOBkheB/AOAtBSBUBOBdkqFEMGs9g9BeACBNAHy1hXQAgwwWR4wNBkhAhJBJAuA9BkgALYxpAnyOBJBXyQLPzOBeBHQLdgL3yHzSA4BYAkB/RoCaNyBKAKKYDEgt4TAYISVhC%2BA6AVVKBYhsLYgIhXxOAeBSAiMBEAB5BgMxbCrAEkEwcQIi0gfAQdJwaER0GSsUakOcHYD81EBobCltWIYic0zwLAbCtEXpPi4ir0JgJ/QOREAEISr5N8/i/gQQEQMQdgKQGQQQRQFQdQGS3QAwRclAGwGwQwPAG8eANYVAeCHIJSjMtCFmcwShKwLgQIE4DMoSrgFK7KKIbKbASwOQAiPCs2Lo1wYRGYfwLgEIYRRYUYGodITIbIAQUqiQWqkoHIKqqoMYcq%2BwIyfoKYDoLwLoJqrqrogYdoNq5YTq3qxqiahYCIEYdqmoNYP8/QUgNEfLUyx858rCmSr8jgVQSMWMDM1CjURcumZi10GKCAH8/ZE4XAbk43L7AwWwqAuiiqAClYICkCtYCAJAWUNIOcainqZ6mjZa7dO65axy4QGNVy6QCGzytQbCgEdw9awwTa0gN8j8naoSucP6uUY%2BPag6o6hi6sCAM6i6zwPpF6/YPQd6laz6tYDkIcMYaKUgcC5IDQFkPYRMSQJC5ISMOC2MQIPYWMFGjC3gFgCQDQC3dG3gHavCpqwirQFYDajgQCtG7C2W2moipW0gPULIFwSQIAA%3D%3D%3D) here also to be sure it it clear enough for everybody.

Reply:
In this, case pure by value parameter would be enough as `unique_ptr` is a move-only type (you cannot copy it, you would get a compile error). 
This is another case for why `unique_ptr` is a good thing. :smile:

Reply:
Thanks @antaljanosbenjamin & @antonio2368!

Reply:
@antonio2368 yes, that's true. The only reason I did say reference, because it is "moved" from one function to another, and it is only used from one place. By using an rvalue reference you can spare that extra move constructor+destructors. This case I would say value parameter has only drawbacks, because the function is never called with an lvalue reference. Though in practice it doesn't matter for performance in my opinion, I think this is the absolute best way to do it.

However if the function would be called with oth lvalue and rvalue references, than the value parameter is the way to go, because it spare and extra overload.